### PR TITLE
DRAFT: Integration of Wasmcov DevHub Proposal

### PR DESCRIPTION
**Pull request by njelich**
This is going to act as my main Pull Request for tracking all the changes needed to implement wasmcov into NEAR Protocol as part of the [Wasmcov Grant](https://near.social/devhub.near/widget/app?page=post&id=2558).

Work is going to need to be done on the near-sdk-rs as well as near-workspaces-rs to fully integrate the novel coverage measurement method.

### CODE REVIEW

**Comment by njelich** [comment]
Taking a look at [near-workspaces-rs set_sandbox_config](https://github.com/near/near-workspaces-rs/blob/51e0ead85556c57f3bd41398cd75025ad51c33b3/workspaces/src/network/config.rs#L60), it would be ideal if the parameters I'm changing here could be set in either the config.json or genesis.json without having to compile nearcore differently (basically if it was a feature of sandbox builds).

@frol @wacban Can you help me figure out the best way to implement this? I see a lot of configs in many different places and I do not want to spend time hacking something together only to completely tear it down. The goal is to call something like the above near-workspaces-rs function to toggle a "no gas limits, no log limits" mode for sandbox builds.

**Comment by wacban** [comment]
Hi @njelich, welcome to Near! 

I'm not familiar with the runtime and wasm but perhaps @nagisa or @Ekleog-NEAR can help you out. That said I think it would be best for you to actually try one or two approaches, put it in a PR and share for review or discussion with concrete questions or concerns.

**Comment by nagisa** [comment]
> The [core parameters files](https://github.com/near/nearcore/tree/master/core/parameters/res) store resource data which is part of the protocol stable enough to be moved outside of the code - can this even be changed in the same build then?

The parameter resources currently are compiled in, but there is not reason why you could not modify the resulting configuration once these are parsed and merged together. In fact we already do so for tests like for example here:

https://github.com/near/nearcore/blob/db20a668b60580229affa8f5af302579a04a7849/core/parameters/src/config.rs#L51-L61

You can very well write

```rust
let config_store = super::config_store::RuntimeConfigStore::new(None); 
let mut current_config = config_store.get_config(PROTOCOL_VERSION);
change_config_however_you_wish(&mut current_config);
invoke_whatever_component_that_consumes_configs(&current_config);
```

That's the basic gist. I’m not sure what your entry point is (are you calling `near-vm-runner::run`? The transaction runtime? something else?) and can’t say whether you have sufficient access to do modification like this at this point in time, though. Alternative could be to add some sort of a function to `RuntimeConfigStore` to register additional diffs. This would perhaps be easier to work with and you’d be able to override configuration at specific protocol versions too.

**Comment by nagisa** [comment]
> All I would do is put the wasmcov_limits config option behind the sandbox flag.

I probably wouldn’t do even that, to be honest. This is not very much new code, it isn’t something that is easy to get wrong and having compile-time features has proven to be more painful than it has been beneficial (at least in my opinion.)

**Comment by njelich** [comment]
@nagisa I have faced  issues with feature flags as predicted - although I can get them to work in a clean repo, I cannot pass the features correctly through to near-sdk-macros (via near-sdk) and to cargo-near (via near-workspaces-rs), so right now the commits in those codebases do not use feautre flags.

I am considering simply using an environment variable, but I'd like to request a review in case I am missing anything.

The other issue I'm facing is a discrepancy between view and transaction calls. I have everything working fine with transactions, but when trying to issue a view call I get the following error:

```
Error: Error { repr: Custom { kind: Rpc(QueryFailure), error: ServerError(HandlerError(ContractExecutionError { vm_error: "wasm execution failed with error: HostError(TotalLogLengthExceeded { length: 274951, limit: 16384 })", block_height: 269262, block_hash: 5KgLpQCpifELhTz7SKwWAmTQNaEcYhF3fGc4FgL13BUP })) } }
```

This indicates that my limit config changes only work for transactions, and not for near-workspace queries. I'd be grateful for a review of that as well.

**To summarize, current blockers are:**
- Wasmcov works with transactions, but not with views, seems to do with limit config not being taken up
- Conditional compilation not passing to 2nd layer of dependencies

**Comment by nagisa** [comment]
> @nagisa I have faced issues with feature flags as predicted - although I can get them to work in a clean repo, I cannot pass the features correctly through to near-sdk-macros (via near-sdk) and to cargo-near (via near-workspaces-rs), so right now the commits in those codebases do not use feautre flags.

I'm not sure I follow -- the PR contains and appears to have no dependencies on feature flags (in the traditional Rust tooling definition of the word.)

But, as a general advise, you can always enable a compile time feature in any transitive dependency by adding a direct dependency on the relevant crates that expose the feature and enable it on that specification. That avoids you having to thread them through `near-sdk` et al. So for example, if `near-sdk` depends on `internal-dependency` and you want to enable a feature of `internal-dependency`, you can write:

```
near-sdk = "1.0"
internal-dependency = { version = "whatever.near.sdk.depends.on", features = [ "whatever-i-need"] }
```

This will enable the `whatever-i-need` feature irrespective of how near-sdk depends on `internal-dependency` due to how Cargo resolves features -- they are strictly additive.

---

> The other issue I'm facing is a discrepancy between view and transaction calls. I have everything working fine with transactions, but when trying to issue a view call I get the following error: <snip>
> 
> This indicates that my limit config changes only work for transactions, and not for near-
workspace queries. I'd be grateful for a review of that as well.

I'm not super familiar what path RPC calls take either. However to me it seems that the entry point for the View calls is the `<ViewClientActor as Handler<WithSpanContext<Query>>>::handle` function. This ultimately uses the `ViewClientActor::config` field. Walking through the callers to figure out how that's set up I get the following chain of callers: 

* `ViewClientActor::new`
* `start_view_client`
* `start_with_config_and_synchronization`
* `RunCmd::run`

It'll be up to you to figure out how to best thread information necessary and at what point to hook in your call to `wasmcov`. That said…

> I am considering simply using an environment variable, but I'd like to request a review in case I am missing anything.

please lets not add yet another new way of configuring the operation of the node. Having at least three distinct mechanisms is already quite daunting.

**Comment by njelich** [comment]
@nagisa The modifications I'm testing with right now are not on the master branch of this PR, since near-workspaces doesnt seem to work with current master.

I linked a commit diff on top of 1.36.5 in my previous comment, but here is a PR for clarity.

https://github.com/near/nearcore/pull/10610

**Comment by akhi3030** [comment]
closing as inactive.
